tools/generate_linux_syscalls.zig

```diff
@@ -48,710 +48,648 @@ fn isReservedNameOld(name: []const u8) bool {
-            .Exited => |code| if (code == 0) child_result.stdout else {
-                std.debug.print("zig cc exited with code {d}\n", .{code});
-                std.process.exit(1);
-            },
-            else => {
-                std.debug.print("zig cc crashed\n", .{});
-                std.process.exit(1);
-            },
-        };
-
-        var lines = mem.tokenizeScalar(u8, defines, '\n');
-        while (lines.next()) |line| {
-            var fields = mem.tokenizeAny(u8, line, " ");
-            const prefix = fields.next() orelse return error.Incomplete;
-
-            if (!mem.eql(u8, prefix, "zigsyscall")) continue;
-
-            const sys_name = fields.next() orelse return error.Incomplete;
-            const value = fields.rest();
-            const name = (getOverridenNameNew(value) orelse sys_name)["sys_".len..];
-            const fixed_name = if (stdlib_renames_new.get(name)) |f| f else if (stdlib_renames.get(name)) |f| f else name;
+fn generateSyscallsFromTable(
+    allocator: std.mem.Allocator,
+    linux_dir: std.fs.Dir,
+    writer: anytype,
+    _arch_info: *const ArchInfo,
+) !void {
+    std.debug.assert(_arch_info.* == .table);
 
-            try writer.print("    {p} = {s},\n", .{ zig.fmtId(fixed_name), value });
-        }
+    const arch_info = _arch_info.table;
 
-        try writer.writeAll("};\n\n");
-    }
-    {
-        try writer.writeAll("pub const RiscV32 = enum(usize) {\n");
-
-        const child_args = [_][]const u8{
-            zig_exe,
-            "cc",
-            "-target",
-            "riscv32-linux-gnuilp32",
-            "-E",
-            "-dD",
-            "-P",
-            "-nostdinc",
-            "-Itools/include",
-            "-Itools/include/uapi",
-            "-D __SYSCALL(nr, nm)=zigsyscall nm nr",
-            "arch/riscv/include/uapi/asm/unistd.h",
-        };
-
-        const child_result = try std.process.Child.run(.{
-            .allocator = allocator,
-            .argv = &child_args,
-            .cwd = linux_path,
-            .cwd_dir = linux_dir,
-        });
-        if (child_result.stderr.len > 0) std.debug.print("{s}\n", .{child_result.stderr});
-
-        const defines = switch (child_result.term) {
-            .Exited => |code| if (code == 0) child_result.stdout else {
-                std.debug.print("zig cc exited with code {d}\n", .{code});
-                std.process.exit(1);
-            },
-            else => {
-                std.debug.print("zig cc crashed\n", .{});
-                std.process.exit(1);
-            },
-        };
-
-        var lines = mem.tokenizeScalar(u8, defines, '\n');
-        while (lines.next()) |line| {
-            var fields = mem.tokenizeAny(u8, line, " ");
-            const prefix = fields.next() orelse return error.Incomplete;
+    // As of 5.17.1, the largest table is 23467 bytes.
+    // 32k should be enough for now.
+    const buf = try allocator.alloc(u8, 1 << 15);
```

Comment:
I did this so that a single allocation could be done up front - and if there were problems one could bump the limit up. Now this gets called for every table arch and never gets `free`'d.

Reply:
This is fixed, buf is just passed now, and `readFile` returns a slice of buf `[0..end]`